runtime/src/iree/hal/drivers/hip/pending_queue_actions.c

```diff
@@ -724,34 +735,46 @@ iree_status_t iree_hal_hip_pending_queue_actions_issue(
     // wait semaphores to make sure that they are either already ready or we can
     // wait on a device event.
     if (action->state == IREE_HAL_HIP_QUEUE_ACTION_STATE_ALIVE) {
-      for (iree_host_size_t i = 0; i < semaphore_count; ++i) {
+      for (iree_host_size_t i = 0; i < action->wait_semaphore_list.count; ++i) {
         // If this semaphore has already signaled past the desired value, we can
         // just ignore it.
         uint64_t value = 0;
         status = iree_hal_semaphore_query(semaphores[i], &value);
         if (IREE_UNLIKELY(!iree_status_is_ok(status))) break;
-        if (value >= values[i]) continue;
-
-        // Try to acquire a hipEvent_t from a device wait timepoint. If so, we
-        // can use that hipEvent_t to wait on the device. Otherwise, this action
-        // is still not ready.
-        hipEvent_t event = NULL;
-        status = iree_hal_hip_event_semaphore_acquire_timepoint_device_wait(
-            semaphores[i], values[i], &event);
-        if (IREE_UNLIKELY(!iree_status_is_ok(status))) break;
-        if (!event) {
+        if (value >= values[i]) {
+          // No need to wait on this timepoint as it has already occurred and
+          // we can remove it from the wait list.
+          iree_hal_semaphore_list_remove_element(&action->wait_semaphore_list,
+                                                 i);
+          --i;
+          continue;
+        }
+
+        // Try to acquire a HIP event from an existing device wait timepoint.
+        // If so, we can use that event to wait on the device.
+        // Otherwise, this action is still not ready for execution.
```

Comment:
Could you also copy and adapt some of your comments in the commit message here so we can have more information of why done this way when reading this part of the codebase?

Reply:
I added some more comments.